### DISCUSSION TITLE
mailer test only works if it's the first one

### DISCUSSION THREAD

**Discussion by yard-mschwartz** [symfony/symfony#57235] [Q&A] [answered]
Symfony version 6.3.12

I'm clueless as to why a certain functional test works in isolation, or if it's the first test in a series of tests, but not when I run the full test suite.

```php
<?php

class EvaluationCreatedEventHandlerTest extends DoctrineTestCase
{
    /** @test */
    public function it_sends_an_email(): void
    {
        // make some objects using zenstruck/foundry factories, including an $evaluatioon

        $event = new EvaluationCreatedEvent($evaluation->getId());

        $handler = static::getContainer()->get(EvaluationCreatedEventHandler::class);
        $handler($event);

        $this->assertQueuedEmailCount(2);
    }
}
```

A `DoctrineTestCase` is just a `KernelTestCase` with a little extra logic so I can use Foundry.

```php
<?php

use Doctrine\ORM\EntityManagerInterface;
use Symfony\Bundle\FrameworkBundle\Test\KernelTestCase;

abstract class DoctrineTestCase extends KernelTestCase
{
    protected EntityManagerInterface $em;

    protected function setUp(): void
    {
        self::bootKernel();

        /** @var EntityManagerInterface $em */
        $em = static::getContainer()->get(EntityManagerInterface::class);

        $this->em = $em;
    }

    protected function tearDown(): void
    {
        parent::tearDown();

        $this->em->clear();
        $this->em->close();
    }
}
```

If I run the full test suite, it claims that no messages were queued. If I change my assertion to `assertEmailCount`, the test doesn't work in either case.

Any idea what I'm missing here?

**Comment by MatTheCat** [comment]
Your `DoctrineTestCase::tearDown` method shouldn’t be necessary: `KernelTestCase` will shut down the kernel itself, so the entity manager will be cleared.

What happens if you remove it?

**Reply by yard-mschwartz** [reply]
It appears I can safely remove these lines. It doesn't have a negative effect on the test suite. It doesn't help for the test I posted in the OP. Still getting the same results.

**Reply by MatTheCat** [reply]
Could you provide a reproducer?

**Reply by yard-mschwartz** [reply]
Haven't had the time do dive deeper myself yet, also because my PHP 8.2 installation has been borked, and I'm unable to reinstall it (using AUR) at the moment.

My gut feeling thinks it may have something to do with our definition of an event bus, namely:

```php
class EventBus implements EventBusInterface
{
    private MessageBusInterface $eventBus;

    public function __construct(MessageBusInterface $eventBus)
    {
        $this->eventBus = $eventBus;
    }

    public function dispatch(object $event): void
    {
        try {
            $envelope = new Envelope($event);

            // tells Symfony to delay execution of this event until after the current process ends
            $envelope->with(new DispatchAfterCurrentBusStamp());

            $this->eventBus->dispatch($envelope);
        }  catch (HandlerFailedException $e) {
            throw ($previous = $e->getPrevious()) ? $previous : $e;
        }
    }
}
```

Perhaps I can try to use a more simple/generic event bus for testing purposes.

If/when I know more, I will certainly post it here. For now, I've marked the test as skipped.

**Reply by MatTheCat** [reply]
`with` methods usually modify and return a copy of the original object, so the `$envelope->with(new DispatchAfterCurrentBusStamp());` line does not serve any purpose.

Don’t think this has to do with your issue though.

**Comment by yard-mschwartz** [comment, marked as answer]
So I was working on some other feature yesterday, and there I also had the problem that tests were working in isolation, but not when I ran all the tests in the file in one go.

Turns out I was missing the `useFactories` trait in my tests, that you need when you work with zenstruck/foundry. More info [here](https://symfony.com/bundles/ZenstruckFoundryBundle/current/index.html#enable-foundry-in-your-testcase). Althought I don't really understand what is happening internally, I do understand you need this trait to ensure tests stay independent of one another. I'm glad the addition of this trait ended up fixing these mailer tests.
